test/tools/ossfuzz/protoToAbiV2.cpp

```diff
@@ -0,0 +1,714 @@
+ * We independently prepare this list for "public" and "external"
+ * callee functions.
+ */
+void ProtoConverter::appendTypedParams(
+	calleeType _calleeType,
+	bool _isValueType,
+	std::string const& _typeString,
+	std::string const& _varName,
+	delimiter _delimiter
+)
+{
+	switch (_calleeType)
+	{
+	case calleeType::PUBLIC:
+		appendTypedParamsPublic(_isValueType, _typeString, _varName, _delimiter);
+		break;
+	case calleeType::EXTERNAL:
+		appendTypedParamsExternal(_isValueType, _typeString, _varName, _delimiter);
+		break;
+	}
+}
+
+// Adds the qualifier "calldata" to non-value parameter of an external function.
+void ProtoConverter::appendTypedParamsExternal(
+	bool _isValueType,
+    std::string const& _typeString,
+    std::string const& _varName,
+    delimiter _delimiter
+)
+{
+	std::string qualifiedTypeString = (
+		_isValueType ?
+		_typeString :
+		_typeString + " calldata"
+	);
+	m_typedParamsExternal << Whiskers(R"(<delimiter><type> <varName>)")
+		("delimiter", delimiterToString(_delimiter))
+		("type", qualifiedTypeString)
+		("varName", _varName)
+		.render();
+}
+
+// Adds the qualifier "memory" to non-value parameter of an external function.
+void ProtoConverter::appendTypedParamsPublic(
+	bool _isValueType,
+	std::string const& _typeString,
+	std::string const& _varName,
+	delimiter _delimiter
+)
+{
+	std::string qualifiedTypeString = (
+		_isValueType ?
+		_typeString :
+		_typeString + " memory"
+		);
+	m_typedParamsPublic << Whiskers(R"(<delimiter><type> <varName>)")
+		("delimiter", delimiterToString(_delimiter))
+		("type", qualifiedTypeString)
+		("varName", _varName)
+		.render();
+}
+
+std::string ProtoConverter::typedParametersAsString(calleeType _calleeType)
+{
+	switch (_calleeType)
+	{
+	case calleeType::PUBLIC:
+		return m_typedParamsPublic.str();
+	case calleeType::EXTERNAL:
+		return m_typedParamsExternal.str();
+	}
+}
+
+// Function that is called by the factory contract
+void ProtoConverter::visit(TestFunction const& _x)
+{
+	m_output << Whiskers(R"(
+	function f() public returns (uint) {
+	)")
+	.render();
+
+	// Define state variables in function scope
+	m_output << m_statebuffer.str();
+
+	// TODO: Support more than one but less than N local variables
+	visit(_x.local_vars());
+
+	m_output << Whiskers(R"(
+		uint returnVal = this.g_public(<parameter_names>);
+		if (returnVal != 0)
+			return returnVal;
+		return (uint(1000) + this.g_external(<parameter_names>));
+	}
+	)")
+	("parameter_names", YulUtilFunctions::suffixedVariableNameList("x_", 0, m_varCounter))
+	.render();
+}
+
+void ProtoConverter::writeHelperFunctions()
+{
+	m_output << Whiskers(R"(
+	function stringCompare(string memory a, string memory b) internal pure returns (bool) {
+		if(bytes(a).length != bytes(b).length)
+			return false;
+		else
+			return keccak256(bytes(a)) == keccak256(bytes(b));
+	}
+	)").render();
+
+	m_output << Whiskers(R"(
+	function bytesCompare(bytes memory a, bytes memory b) internal pure returns (bool) {
```

Comment:
Why use a different way of comparing for bytes an string?

Reply:
you are right. I could use a single function and cast string to bytes or vice versa in the function parameter when calling the unified compare function. Do you have any suggestions?

Reply:
I can't say which version would be faster or better to use.